query/src/storages/fuse/table_functions/fuse_history_truncate.rs

```diff
@@ -0,0 +1,275 @@
+        let prev_blocks: HashSet<String> = self.blocks_of(da.clone(), seg_delta.iter()).await?;
+        let current_blocks: HashSet<String> =
+            self.blocks_of(da.clone(), current_segments.iter()).await?;
+        let block_delta = prev_blocks.difference(&current_blocks);
+
+        // NOTE: the following actions are NOT transactional yet
+
+        // 1. remove blocks
+        let mut block_removed = 0u64;
+        for x in block_delta {
+            self.remove_location(da.clone(), x).await?;
+            block_removed += 1;
+        }
+
+        // 2. remove the segments
+        let mut segment_removed = 0u64;
+        for x in seg_delta {
+            self.remove_location(da.clone(), x).await?;
+            segment_removed += 1;
+        }
+
+        // 3. remove the snapshots
+        for x in snapshots.iter().rev() {
+            let loc = snapshot_location(&x.snapshot_id);
+            self.remove_location(da.clone(), loc).await?
+        }
+
+        let snapshot_removed = snapshots.len() as u64;
+
+        Ok(Some((snapshot_removed, segment_removed, block_removed)))
+    }
+
+    fn empty_result(&self) -> Result<SendableDataBlockStream> {
+        self.build_result(vec![])
+    }
+
+    fn build_result(&self, blocks: Vec<DataBlock>) -> Result<SendableDataBlockStream> {
+        Ok(Box::pin(DataBlockStream::create(
+            self.table_info.schema(),
+            None,
+            blocks,
+        )))
+    }
+
+    async fn blocks_of(
+        &self,
+        data_accessor: Arc<dyn DataAccessor>,
+        locations: impl Iterator<Item = impl AsRef<str>>,
+    ) -> Result<HashSet<String>> {
+        let mut result = HashSet::new();
+        for x in locations {
+            let res: SegmentInfo = io::read_obj(data_accessor.as_ref(), x).await?;
+            for block_meta in res.blocks {
+                result.insert(block_meta.location.path);
+            }
+        }
+        Ok(result)
+    }
+
+    async fn remove_location(
+        &self,
+        data_accessor: Arc<dyn DataAccessor>,
+        location: impl AsRef<str>,
+    ) -> Result<()> {
+        data_accessor.remove(location.as_ref()).await
+    }
+}
+
+#[async_trait::async_trait]
+impl Table for FuseTruncateHistory {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn get_table_info(&self) -> &TableInfo {
+        &self.table_info
+    }
+
+    fn table_args(&self) -> Option<Vec<Expression>> {
+        let flag_truncate_all = if self.truncate_all { "all" } else { "" };
+        Some(vec![
+            string_literal(self.arg_database_name.as_str()),
+            string_literal(self.arg_table_name.as_str()),
+            string_literal(flag_truncate_all),
+        ])
+    }
+
+    async fn read(
+        &self,
+        ctx: Arc<QueryContext>,
+        _plan: &ReadDataSourcePlan,
+    ) -> Result<SendableDataBlockStream> {
+        if let Some((snapshot_removed, segment_removed, block_removed)) =
+            self.truncate_history(ctx).await?
```

Comment:
Why use the read trait to do the truncate operation? 

`select * from fuse_truncate_history('db_09_0007', 't');` --> `truncate table fuse_truncate_history('db_09_0007', 't') ` ?

Reply:
or `fuse_truncate_history('db_09_0007', 't')` ?

Reply:
`truncate table fuse_truncate_history('db_09_0007', 't')` is likely to be understood as table `fuse_truncate_history` is going to be truncated.

`call fuse_truncate_history('db_09_0007', 't')` might be better?

Reply:
```
mysql> select * from fuse_history('default', 't1');
+----------------------------------+----------------------------------+---------------+-------------+-----------+--------------------+------------------+
| snapshot_id                      | prev_snapshot_id                 | segment_count | block_count | row_count | bytes_uncompressed | bytes_compressed |
+----------------------------------+----------------------------------+---------------+-------------+-----------+--------------------+------------------+
| fe6076fa4c444fd5b7089585c24f8025 | 9d95a99799d24d8984c41372e4b674f4 |             4 |           4 |         5 |                 20 |              775 |
| 9d95a99799d24d8984c41372e4b674f4 | 7cdd8c0aee364fd2a727eca6f0c506da |             3 |           3 |         4 |                 16 |              583 |
| 7cdd8c0aee364fd2a727eca6f0c506da | c158653e18f241979fc250facf1091bd |             2 |           2 |         3 |                 12 |              391 |
| c158653e18f241979fc250facf1091bd | NULL                             |             1 |           1 |         2 |                  8 |              199 |
+----------------------------------+----------------------------------+---------------+-------------+-----------+--------------------+------------------+
4 rows in set (0.02 sec)
Read 4 rows, 465 B in 0.013 sec., 311.81 rows/sec., 36.25 KB/sec.

mysql> select * from fuse_truncate_history('default', 't1'); -- this likes truncate table fuse_history('default', 't1');
+------------------+-----------------+---------------+
| snapshot_removed | segment_removed | block_removed |
+------------------+-----------------+---------------+
|                3 |               0 |             0 |
+------------------+-----------------+---------------+
1 row in set (0.03 sec)
Read 1 rows, 24 B in 0.023 sec., 44.34 rows/sec., 1.06 KB/sec.

mysql> select * from fuse_history('default', 't1');
+----------------------------------+----------------------------------+---------------+-------------+-----------+--------------------+------------------+
| snapshot_id                      | prev_snapshot_id                 | segment_count | block_count | row_count | bytes_uncompressed | bytes_compressed |
+----------------------------------+----------------------------------+---------------+-------------+-----------+--------------------+------------------+
| fe6076fa4c444fd5b7089585c24f8025 | 9d95a99799d24d8984c41372e4b674f4 |             4 |           4 |         5 |                 20 |              775 |
+----------------------------------+----------------------------------+---------------+-------------+-----------+--------------------+------------------+
1 row in set (0.03 sec)
Read 1 rows, 136 B in 0.013 sec., 74.98 rows/sec., 10.2 KB/sec.

mysql> select * from fuse_truncate_history('default', 't1');
Query OK, 0 rows affected (0.02 sec)
Read 0 rows, 0 B in 0.010 sec., 0 rows/sec., 0 B/sec.

```

Not sure `fuse_truncate_history` table function is for? It seems `fuse_history` table can do all the operations what we did with the `fuse_truncate_history`.
Thanks.

Reply:
`fuse_truncate_history` is a function with side effects, which
 
 -  truncate the historical data of the given table ( or kinda of GC)
     for an append-only table, like `t1` you given in the example, only  snapshot file(on disk/s3) will be removed, since the table is append-only, all the segments and blocks are still in use.
 - returns the number of snapshot/segment/block that has been removed.    
    
 try `insert overwrite t1 values(...)`, and then `select * from fuse_truncate_history(...)`.  here is an example:
 
 ~~~
 create table t(c int);
 insert into t select * from numbers(100);
 ~~~
 after the insertion , the storage dir may look like:
 
 ~~~
 stateless_test_data/
├── _b
│   ├── 4bdb25420b7949058b55c067028674d6.parquet
│   ├── 6c8375b2309a4a918e5a6c52bc0c4fb2.parquet
│   ├── ....
│   └── fda4f29d751f4c8d8afc93656f7dfeb8.parquet
├── _sg
│   ├── 051beb7eb7214276896ce446d52acc5d
│   ├── 27cba26cc2954315875f26d14aede9a5
│   ├── ...
│   └── f55416c3c6d74c1ea958f1a19a472bff
└── _ss
    └── 36d794938e12450f814c069c8df0ee33
 ~~~
 
 and then, we insert **OVERWRITE** and truncate
 
 ~~~
 insert overwrite t values(1);
select * from fuse_truncate_history('default', 't');
 ~~~   
 after that
 
 ~~~
stateless_test_data/
├── _b
│   └── f04eedbd25b14ab8b2dd60feb68cf299.parquet
├── _sg
│   └── ebb24d9071f34f1fab3a5981e9413602
└── _ss
    └── a1d51110980f4c9ea23faa3398d5bc49
 ~~~
 
 only the data that referenced by the last snapshot are kept
 
---------------



as stated in the summary, it is not intended to be used by the end user.

the front-end syntax has not been settled,  previously, I am planning to expose this functionality as `call` statement  (in other PR),  but any other suggestions are welcome.

the `select * from func` syntax is used by `interpreter_drop`, we could use `FuseTruncateHistory` directly in the interpreter, but exposing a fuse table specific concept into the interpreters module may not be a good idea.

Reply:
> Why use the read trait to do the truncate operation?

forget the answer this :) :

- ideally, the GC process should be executed in a distributed manner. 
   the `read_plan`  and then distributedly  `read` mechanism is too tempting
-  promised  @wubx to provide a preview of `drop/truncate` with GC before Fri.
   unfortunately, seem there is not enough time to deliver a distributed version before Fri.
      
there must better ideas, pls let me know